### PR TITLE
Refactor strmatcher.MphMatcherGroup

### PR DESCRIPTION
**Pull request by Vigilans**
As per description in https://github.com/v2fly/v2ray-core/discussions/1275#discussioncomment-1559883:

> `MphMatcherGroup` uses an algorithm [Hash, displace, and compress](http://cmph.sourceforge.net/papers/esa09.pdf) to construct a minimal perfect hash table. It involves `1. a normal hash function`, and `2. a series of hash functions independent to each others`.
> 
> This is how this `MphMatcherGroup` works:
> 
> #### First level hash: RollingHash
> Rolling hash is chosen here because we can easily calculate the hash of a full domain and all its subdomains.
> 
> We obtain a seed from the result of first level hash.
> 
> #### Second level hash: MemHash
> One of second level functions is selected according to the seed obtained from first level hash.
> 
> This is done by choosing golang runtime's `memhash`, because `memhash` requires a seed as argument, and with different seeds, `memhash` behaves exactly like distinct hash functions independent to each other.
> 
> To use `memhash`, current implementation of `MphMatcherGroup` copies source code (unoptimized version) from golang runtime. However, we could directly reference the runtime 'simplementation using compiler directive `go:linkname`:
> 
> ```
> //go:noescape
> //go:linkname memhash runtime.memhash
> func memhash(p unsafe.Pointer, h, s uintptr) uintptr
> ```
> 
> While avoiding duplicate code, the runtime's memhash implementation could also benefit from hardware optimizations like AES if supported. Current `MphMatcherGroup` unconditionally uses the unoptimized fallback version.

### CODE REVIEW

File: common/strmatcher/matchergroup_mph.go [hunk 1]
```diff
@@ -149,109 +202,6 @@ func nextPow2(v int) int {
...
-	return (x << 31) | (x >> (64 - 31))
-}
-
-func readUnaligned64(p unsafe.Pointer) uint64 {
-	q := (*[8]byte)(p)
-	return uint64(q[0]) | uint64(q[1])<<8 | uint64(q[2])<<16 | uint64(q[3])<<24 | uint64(q[4])<<32 | uint64(q[5])<<40 | uint64(q[6])<<48 | uint64(q[7])<<56
-}
+//go:noescape
+//go:linkname strhash runtime.strhash
+func strhash(p unsafe.Pointer, h uintptr) uintptr
```
Comment (Vigilans) [inline common/strmatcher/matchergroup_mph.go:207]: `memhash`'s source code is removed, and directly referenced using golang's compiler directive `//go:linkname`.

The removed source code above is a fallback unoptimized version, could not benefit hardware optimizations like AES.

`strhash` is a wrapper of `memhash` in golang runtime to compute the memhash for a string.

File: common/strmatcher/matchergroup_mph_test.go [hunk 2]
```diff
@@ -172,3 +173,97 @@ func TestMphMatcherGroup(t *testing.T) {
 		}
 	}
 }
+
+// See https://github.com/v2fly/v2ray-core/issues/92#issuecomment-673238489
+func TestMphMatcherGroupAsIndexMatcher(t *testing.T) {
```
Comment (Vigilans) [inline common/strmatcher/matchergroup_mph_test.go:178]: This test views `MphMatcherGroup` as an `IndexMatcher` (without regex and substr matching capability), and use tests from `LinearIndexMatcher` to prove it has the capability to serve as matcher for DNS module.

**Comment by Vigilans** [comment]
### Benchmark

#### Domain Matcher

* Original `MatchAny`: 34.95 ns/op
* New `MatchAny`: 32.49 ns/op
* New `Match`: 40.25 ns/op

#### Full Matcher

* Original `MatchAny`: 34.67 ns/op
* New `MatchAny`: 26.86 ns/op
* New `Match`: 40.43 ns/op

Because only Domain matcher needs to lookup the mph table every time a dot '.' is met, so `ACAutomationMatcherGroup` could actually be further optimized according to what matchers are stored in it and what type the upstream wants to match. This will be implemented in `EtldIndexMatcher`'s PR.

**Review by darsvador — COMMENTED** [review]
````go
        keyLen := len(*g.ruleMap)
	if keyLen == 0 {
		keyLen = 1
		(*g.ruleMap)["empty___"] = RollingHash("empty___")
	}
````
I noticed that you deleted the empty rule code when `ruleCount` is 0. Can [this situation](https://github.com/v2fly/v2ray-core/issues/785) be handled properly now?

File: common/strmatcher/matchergroup_mph.go [hunk 3]
```diff
@@ -10,134 +10,187 @@ import (
...
+	info.matchers[matcherType] = append(info.matchers[matcherType], value)
+	(*g.ruleInfos)[fullPattern] = info
+	return info.rollingHash
+}
+
+// Build builds a minimal perfect hash table for insert rules.
+// Algorithm used: Hash, displace, and compress. See http://cmph.sourceforge.net/papers/esa09.pdf
+func (g *MphMatcherGroup) Build() error {
+	ruleCount := len(*g.ruleInfos)
+	g.level0 = make([]uint32, nextPow2(ruleCount/4))
```
[the remark this answers is not in the document]
Comment (Vigilans) [reply common/strmatcher/matchergroup_mph.go:91]: The index 0 is reserved for failed lookup now:
https://github.com/v2fly/v2ray-core/blob/babac85df007c8416587c6d781992900e5400e39/common/strmatcher/matchergroup_mph.go#L39-L41
 and filled with dummy value ("" and nil) at the creation of `MphMatcherGroup`:
https://github.com/v2fly/v2ray-core/blob/babac85df007c8416587c6d781992900e5400e39/common/strmatcher/matchergroup_mph.go#L49-L52

When ruleCount is 0, `nextPow2(0)` will return 1, so `g.level0` and `g.level1` are guaranteed to be created with length 1, and corresponding masks to be 0:

https://github.com/v2fly/v2ray-core/blob/babac85df007c8416587c6d781992900e5400e39/common/strmatcher/matchergroup_mph.go#L90-L94

Therefore, during `Lookup(hash, input)`, both `i0`, `seed`, `i1`, `n` will all be 0, so the returned index is 0, indicating failed lookup:
https://github.com/v2fly/v2ray-core/blob/babac85df007c8416587c6d781992900e5400e39/common/strmatcher/matchergroup_mph.go#L143-L147

[the remark this answers is not in the document]
Comment (Vigilans) [reply common/strmatcher/matchergroup_mph.go:91]: @darsvador But you're right, I should add a testcase to test MphMatcherGroup could safely work with empty matchers.

**Review by darsvador — APPROVED** [review]

**Review by xiaokangwang — APPROVED** [review]
This pull request does not contains changes that can break other part of V2. I am not an expert on this matching methods so it is advised for other members to check the matching part.

If there is no other issue subsequently discovered, this pull request can be merged.
